Sources/SMTPKitten/Connection/SMTPClient.swift

```diff
@@ -0,0 +1,157 @@
+            requestWriter.yield(request)
+        }
+    }
+
+    internal func run() async throws {
+        do {
+            try await channel.executeThenClose { inbound, outbound in
+                var inboundIterator = inbound.makeAsyncIterator()
+
+                for await request in requests {
+                    do {
+                        if request.buffer.readableBytes > 0 {
+                            // The first "message" on a connection send by us is empty
+                            // Because we're expecting to read data here, not write
+                            try await outbound.write(request.buffer)
+                        }
+
+                        guard var lastLine = try await inboundIterator.next() else {
+                            throw SMTPClientError.endOfStream
+                        }
+
+                        let code = lastLine.code
+                        var lines = [lastLine]
+
+                        while !lastLine.isLast, let nextLine = try await inboundIterator.next() {
+                            guard nextLine.code == code else {
+                                throw SMTPClientError.protocolError
+                            }
+
+                            lines.append(nextLine)
+                            lastLine = nextLine
+                        }
+
+                        request.continuation.resume(
+                            returning: SMTPReply(
+                                code: code,
+                                lines: lines.map(\.contents)
+                            )
+                        )
+                    } catch {
+                        request.continuation.resume(throwing: error)
+                        throw error
+                    }
+                }
+            }
+
+            requestWriter.finish()
+            for await request in requests {
+                request.continuation.resume(throwing: SMTPClientError.endOfStream)
+            }
+        } catch {
+            self.error = error
+            requestWriter.finish()
+            for await request in requests {
+                request.continuation.resume(throwing: error)
+            }
+            throw error
+        }
+    }
+
+    nonisolated fileprivate func starttls(
+        configuration: SMTPSSLConfiguration,
+        hostname: String
+    ) async throws {
+        try await send(.starttls)
+            .status(.serviceReady, or: SMTPClientError.startTLSFailure)
+
+        let sslContext = try NIOSSLContext(configuration: configuration.configuration.makeTlsConfiguration())
+        let sslHandler = try NIOSSLClientHandler(context: sslContext, serverHostname: hostname)
+
+        try await self.channel.channel.pipeline.addHandler(sslHandler, position: .first)
+    }
+
+    public static func withConnection<T>(
+        to host: String,
+        port: Int = 587,
+        ssl: SMTPSSLMode,
+        perform: (SMTPClient) async throws -> T
```

Comment:
Given the only function that is not private in `SMTPClient` is `send` (if you make `run` private) you could split off the member variables you need to implement `send` (ie the continuation) into a separate type which is easier to make Sendable.

Although you also have the variable `handshake` but that seems to only be accessed internally